lib/block_view_atomic_txns_test.go

```diff
@@ -0,0 +1,667 @@
+	require.NoError(t, err)
+	atomicTxnsWrapperDuplicate.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns[2] =
+		atomicTxnsWrapperDuplicate.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns[1]
+	atomicTxnsWrapperDuplicate.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns[1] = txnCopy
+	_, err = _atomicTransactionsWrapperWithConnectTimestamp(
+		t, testMeta.chain, testMeta.db, testMeta.params, atomicTxnsWrapperDuplicate, 0)
+	require.Contains(t, err.Error(), RuleErrorAtomicTxnsHasBrokenChain)
+
+}
+
+func TestDependentAtomicTransactionGeneration(t *testing.T) {
+	// Initialize test chain, miner, and testMeta.
+	testMeta := _setUpMinerAndTestMetaForAtomicTransactionTests(t)
+
+	// Initialize m0, m1, m2, m3, m4.
+	_setUpUsersForAtomicTransactionsTesting(testMeta)
+
+	// Generate 100 dependent atomic transactions.
+	atomicTxns, _ := _generateSignedDependentAtomicTransactions(testMeta, 100)
+
+	// Construct a new view to connect the transactions to.
+	utxoView, err := NewUtxoView(
+		testMeta.db, testMeta.params, testMeta.chain.postgres, testMeta.chain.snapshot, nil)
+	require.NoError(t, err)
+	blockHeight := testMeta.chain.BlockTip().Height + 1
+
+	// Get the initial balance for m0.
+	m0InitialBalanceNanos := _getBalance(t, testMeta.chain, testMeta.mempool, m0Pub)
+
+	// Connect the transactions to ensure they can actually be connected.
+	var totalFees uint64
+	for _, txn := range atomicTxns {
+		// Connect the transaction.
+		txHash := txn.Hash()
+		_, _, _, fees, err := utxoView.ConnectTransaction(
+			txn, txHash, blockHeight, 0, true, false)
+		require.NoError(t, err)
+		totalFees += fees
+	}
+
+	// Flush the view to ensure everything is working properly.
+	require.NoError(t, utxoView.FlushToDb(uint64(blockHeight)))
+
+	// Get the final balance for m0.
+	m0FinalBalanceNanos := _getBalance(t, testMeta.chain, testMeta.mempool, m0Pub)
+
+	// Check that fees were paid.
+	require.Equal(t, m0InitialBalanceNanos-totalFees, m0FinalBalanceNanos)
+
+	//
+	// Now we test that the transactions are truly dependent on each-other by reorganizing them.
+	//
+
+	// Generate 100 new dependent atomic transactions.
+	atomicTxns, _ = _generateSignedDependentAtomicTransactions(testMeta, 100)
+
+	// Reorganize the transactions.
+	initialTxn := atomicTxns[0]
+	atomicTxns[0] = atomicTxns[len(atomicTxns)-1]
+	atomicTxns[0] = initialTxn
+
+	// Construct a new view to connect the transactions to.
+	utxoView, err = NewUtxoView(
+		testMeta.db, testMeta.params, testMeta.chain.postgres, testMeta.chain.snapshot, nil)
+	require.NoError(t, err)
+	blockHeight = testMeta.chain.BlockTip().Height + 1
+
+	// Connect the transactions to ensure they can actually be connected.
+	for _, txn := range atomicTxns {
+		// Connect the transaction.
+		txHash := txn.Hash()
+		_, _, _, _, err := utxoView.ConnectTransaction(
+			txn, txHash, blockHeight, 0, true, false)
+		if err != nil {
+			require.Contains(t, err.Error(), RuleErrorInsufficientBalance)
+		}
+	}
```

Comment:
can we add a call to `_executeAllTestRollbackAndFlush` to make sure all connect and disconnect stuff behaves as expected please?